### DISCUSSION TITLE
VSCode integration no longer working

### DISCUSSION THREAD

**Discussion by falkben** [akaihola/darker#516] [Q&A]
Some teammates are reporting that darker is no longer functioning as a formatting provider inside vscode.

The options suggested in the readme are currently deprecated:

```json
"python.formatting.provider": "black",
"python.formatting.blackPath": "<install_location_from_step_2>",
"python.formatting.blackArgs": [],
```

It seems that the vscode Python extension is moving most of it's previously included integrations (like black formatting provider) into separate extensions. They have documentation on this transition here: https://github.com/microsoft/vscode-python/wiki/Migration-to-Python-Tools-Extensions 

I am wondering if a new `darker` extension will be needed or if darker could possible tag along and run inside the black extension, similar to how it's been running as the black formatting provider. I noticed that there is a GH issue for this, but unsure whether it will actually be needed: https://github.com/akaihola/darker/issues/424  Would you need help with this?

In the meantime, should the README be updated to include some information about the feasibility of running inside vscode during this transition period?

**Comment by falkben** [comment]
After some experimentation on my end, I'm pretty sure we cannot currently use the [black extension](https://marketplace.visualstudio.com/items?itemName=ms-python.black-formatter), as is, to run `darker`.

If the black extension is installed, it becomes the default formatter, but will use the python.formatting.blackPath as configured. However, it fails to format, as it does a version check at startup, and the darker version (`1.7.2`) is too low, compared to the black version it currently checks for `22.3.0`.

If you try to explicitly configure the black extension to use the darker path through the black extension config:

```json
 "black-formatter.path": "<install_location_from_step_2>",
```

then the black extension fails to load at startup

**Reply by kedhammar** [reply]
I've made a related issue in #525 

Any progress on this or ideas how to handle the issue?

**Reply by Ashblaze** [reply]
Please check my comment below, it might help.

**Comment by Ashblaze** [comment]
As a workaround, it is possible to use "[Run on Save](https://marketplace.visualstudio.com/items?itemName=emeraldwalk.RunOnSave)" or "emeraldwalk" extension from the VSCode Marketplace to automatically trigger darker.

**Steps:**
1. Set "Default Formatter" as "None" in your VS Code settings.
2. Uncheck "Format on Save" (OR) remove/disable Black (and all other) formatter extension from VS Code.
3. Install extension "Run on Save" (emeraldwalk) from marketplace.
4. Add darker executable's folder location to system path.
5. Add the following to settings.json (you can modify this to suit your needs):
```
"emeraldwalk.runonsave": {
    "commands": [
        {
            "match": ".py",
            "cmd": "darker \"${file}\" -S"
        }
    ]
}
```

**Comment by akaihola** [comment]
Thanks @falkben, @kedhammar and @Ashblaze for your report and analysis. I'm trying to form a picture of the current status with regard to interoperability between Darker and VSCode. Any updates to the situation since August 2023?
